Alcohol Refactor (Merge-Ready)

Xhuis opened the pull request:
:cl: Xhuis
experiment: Alcohol has received multiple tweaks.
tweak: Drunkenness will now be displayed on examine as long as the target's face is not hidden, ranging from a slight flush to being a completely drunken wreck. This also helps to measure alcohol poisoning, although the appearance will remain for thirty seconds after the alcohol leaves the imbiber's bloodstream.
tweak: Alcohol poisoning now has different effects. It begins with standard things such as confusion, dizziness, and slurring, but eventually escalates to constant vomiting, blacking out, and toxin damage.
tweak: Drinks now have a wide variety of alcoholism. Grog is barely a drink whatsoever, for instance, but Manly Dorf will completely floor you with too much consumption. Note that drinks based off of whiskey and vodka are the hardest, whereas beer and wine are the softest.
tweak: The speed of alcohol poisoning has been slightly reduced.
rscadd: Some drinks have received unique effects. Experiment!
rscadd: A new base alcohol joins the lineup! Absinthe has been added, and is very strong.
rscadd: Whiskey Sour, made with Whiskey, Lemon Juice, and Sugar, simply expands the existing whiskey-based lineup.
rscadd: Fetching Fizz, made with Nuka-Cola and Iron, will pull all nearby ores in the direction of the imbiber.
rscadd: Hearty Punch, made with Brave Bull, Syndicate Bomb, and Absinthe, will provide extreme healing in critical condition.
rscadd: Bacchus' Blessing, made with four of the most powerful alcohols on the station, is over three times stronger than any other alcohol on the station. Consume in small amounts or risk death.
/:cl:

On the backend, boozepwr has been changed to cause more severe effects with a higher number (as it should have been in the first place!) and now has more diverse effects depending on different factors - primarily, the strongness of the drink and how long it's been in the imbiber's bloodstream.

**To-Do**
- [x] More drinks (from #16134)
- [x] Unique effects for certain drinks
### And no, I'm not back! I did this because I was bored and needed the practice.

Commit:
Alcohol refactor

Xhuis commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -19,36 +8,91 @@
-			. = 1
+	if(M.alcohol_tolerance == COMPLETE_TOLERANCE)
+		return ..() || .
+	var/alcohol_effect = (boozepwr * (current_cycle / 50)) / M.alcohol_tolerance
```
Tried to use a switch() here but it didn't want to work correctly so this is probably awful.

RemieRichards replied:
```
switch(alcohol_effect)
    if(1 to 10)
        ...
    if(11 to 20)
        ...
    if(21 to 40)
        ...
```

Xhuis replied:
I want the drunkenness effects from earlier stages to carry over. Is that possible?

Iamgoofball commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -58,20 +102,34 @@
+	if(iscarbon(M))
+		var/mob/living/carbon/human/H = M
+		spawn(300) //30 seconds
+			H.drunk_state = NOT_DRUNK //Shouldn't be a problem if they have more alcohol in their system since it'll immediately be set back
```
should probably implement hangover

lordpidey replied:
If someone drinks two drinks, with different quantities, wouldn't this cause drunkeness to reset after the first one metabolizes?

Iamgoofball commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -329,37 +387,37 @@
 	description = "Beer and Ale, brought together in a delicious mix. Intended for true men only."
 	color = "#664300" // rgb: 102, 67, 0
-	boozepwr = 45 //was 10, but really its only beer and ale, both weak alchoholic beverages
+	boozepwr = 100 //was 10, but really its only beer and ale, both weak alchoholic beverages //FUCK THAT SHIT, THIS IS FOR MANLY MEN
```
*manly dorfs

KazeEspada replied:
Isnt 100 deadly?

Xhuis replied:
In small doses, no.

ChangelingRain commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -58,20 +102,34 @@
+/datum/reagent/consumable/ethanol/on_mob_delete(mob/living/M)
+	if(iscarbon(M))
+		var/mob/living/carbon/human/H = M
+		spawn(300) //30 seconds
```
Make sure the mob still exists at this point.

Xhuis replied:
Will do

Zilenan91 commented:
The effect you added to one of them is pretty neat.

PKPenguin321 commented:
yayyy, xhuis is back forever and ever! :tada:

Zilenan91 commented:
forever and ever tgstation.org forever and ever Xhuis tgstation.org

PKPenguin321 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -19,36 +8,91 @@
+		M.drunk_state = BARELY_ALIVE
+
+	if(alcohol_effect >= 101)
+		M.adjustToxLoss(boozepwr) //Let's be honest you shouldn't be alive by now
```
does this mean a mixture of manly dorf and hooch instakills you

Xhuis replied:
Pretty quickly it does yeah, but the effect needs time to stack on both drinks so it wouldn't unless there was maybe 15-20 units of each

Xhuis commented:
New in this commit:
* Alcohol is no longer handled via the reagent, but by the new proc "handle_alcohol()" in the life.dm for humans
* Humans have four new variables: drunkenness, which measures overall drunktitude; drunk_state, which is used for examining; alcohol_tolerance, which is self-explanatory; and hungover, which is not yet implemented.
* Drunkenness from multiple drinks now stacks - two extremely potent drinks will stack in terms of how fast you will become drunk
* When at high alcohol levels, you have a high chance to pass out (unless the shuttle is currently docked) and will not progress to the final stage of alcohol poisoning when asleep
* revive() now removes drunkenness from humans

Iamgoofball commented:
>Alcohol is no longer handled via the reagent, but by the new proc "handle_alcohol()" in the life.dm for humans

:-1: :x: :100:

this goes against how reagents are supposed to be used, all reagent effects need to be self contained

Xhuis commented:
The reagents still play a key part in making someone drunk, and the drinks still have unique effects. I just didn't want to have a large on_mob_life() for ethanol primarily, plus this allows for more flexibility.

lordpidey commented on code/modules/mob/living/living_defines.dm:
```diff
@@ -55,6 +70,6 @@
 
 	var/last_bumped = 0
-	var/unique_name = 0 //if a mob's name should be appended with an id when created e.g. Mob (666)
+	var/unique_name = 0 //if a mob's name should be appended with an id when created e.g. Mob (69)
```
Don't change that, mobs have a THREE digit ID, your comment implies that IDs are 2 digit.

Xhuis replied:
:^)

phil235 replied:
yeah, please don't change the comment.

Xhuis commented:
New in this commit:
* Anatomic Panacea now purges drunkenness
* Antihol now purges drunkenness
* Alcohol poisoning is now less lethal in the final two stages
* Hooch now reduces alcohol tolerance
* Cuba Libre now provides minor healing to revolutionaries
* Screwdriver Cocktail now massively reduces radiation on engineering staff
* Brave Bull now grants +5 maximum health while in your system
* Manly Dorf's sprite no longer has a block of white pixels between the handle and mug
* Barefoot now quickly heals brute damage if the imbiber is not wearing shoes
* Demon's Blood and Devil's Kiss now cause all nearby slaughter demons to leech health from the imbiber
* Syndicate Bombs now have a small chance to cause an explosion sound at the imbiber's location
* Green Beer now turns the imbiber green while it's in their system

lordpidey commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -180,35 +193,48 @@
+/datum/reagent/consumable/ethanol/hooch/on_mob_delete(mob/living/M)
+	if(ishuman(M))
+		var/mob/living/carbon/human/H = M
+		if(H.alcohol_tolerance != initial(H.alcohol_tolerance))
```
Did you mean <= instead of !=  ?

Xhuis replied:
Nope

ChangelingRain commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -422,30 +470,43 @@
+
+/datum/reagent/consumable/ethanol/demonsblood/on_mob_life(mob/living/M)
+	for(var/mob/living/simple_animal/slaughter/S in range(10, M)) //Demon's Blood causes nearby slaughter demons to sap the life from the imbiber.
+		S.adjustBruteLoss(1)
```
S.adjustHealth(-1)

Xhuis replied:
adjustHealth isn't a proc.

ChangelingRain commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -481,18 +542,24 @@
+
+/datum/reagent/consumable/ethanol/devilskiss/on_mob_life(mob/living/M)
+	for(var/mob/living/simple_animal/slaughter/S in range(10, M)) //Devil's Kiss has the same effect as Demon's Blood, only stronger.
+		S.adjustBruteLoss(2)
```
like the above comment but with -2

Supermichael777 commented:
>Cuba Libre now provides minor healing to revolutionaries

VALIDBONER ACTIVATE!

PKPenguin321 commented:
not like you can't just beat a rev in the head a couple times to not only confirm they're a rev but deconvert them in the process

KazeEspada commented:
TIME TO REV TEST PEOPLE WITH CUBA LIBRES

GunHog commented:
>Manly Dorf will completely floor you with too much consumption.

100% Magni Bronzebeard approved!

Cheridan commented:
>Cuba Libre now provides minor healing to revolutionaries

this kind of thing has proven in the past to be an awful idea but yeah in this case if you have someone tied up and you think they're a rev you're probably going to either implant them or beat them

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -222,56 +248,69 @@
+		M.adjustBruteLoss(-1)
+		M.adjustFireLoss(-1)
+		M.adjustToxLoss(-1)
+		M.adjustOxyLoss(-5)
```
need a 0 as second argument for each of those.

Xhuis replied:
I'll fix that. What's the difference between `..()` and `..() || .`?

phil235 replied:
`.` is what the current proc returns.
`..()` is calling the parent proc.
We want the proc to return 1 when any of the adjustxxxx() is called, so if there's one in the current proc we do `. = 1`, but we also have to call `..()` because we want to run the code in the parent proc. This parent proc itself can have an adjustxxx(), so to cover all cases we need `return ..() || .` 
Sometimes the current proc doesn't have an adjustxxx() but the parent will have one, so just doing `..()` won't make our proc return the needed 1.

Xhuis replied:
Ah, okay, thank you

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -222,56 +248,69 @@
+/datum/reagent/consumable/ethanol/screwdrivercocktail/on_mob_life(mob/living/M)
+	if(M.mind && M.mind.assigned_role in list("Station Engineer", "Atmospheric Technician", "Chief Engineer")) //Engineers lose radiation poisoning at a massive rate.
+		M.radiation = max(M.radiation - 25, 0)
+	..()
```
return ..()

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -285,28 +324,38 @@
+/datum/reagent/consumable/ethanol/brave_bull/on_mob_life(mob/living/M)
+	if(M.maxHealth == initial(M.maxHealth)) //Brave Bull makes you sturdier, and thus capable of withstanding a tiny bit more punishment.
+		M.maxHealth += 5
+	..()
```
return ..()

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -285,28 +324,38 @@
+/datum/reagent/consumable/ethanol/brave_bull/on_mob_delete(mob/living/M)
+	if(M.maxHealth != initial(M.maxHealth))
+		M.maxHealth = initial(M.maxHealth)
+	..()
```
return ..()

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -422,30 +470,43 @@
+	if(ishuman(M)) //Barefoot causes the imbiber to quickly regenerate brute trauma if they're not wearing shoes.
+		var/mob/living/carbon/human/H = M
+		if(!H.shoes)
+			H.adjustBruteLoss(-3)
```
```
H.adjustBruteLoss(-3)
. = 1
```

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -422,30 +470,43 @@
+		var/mob/living/carbon/human/H = M
+		if(!H.shoes)
+			H.adjustBruteLoss(-3)
+	..()
```
return ..() || .

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -422,30 +470,43 @@
+	for(var/mob/living/simple_animal/slaughter/S in range(10, M)) //Demon's Blood causes nearby slaughter demons to sap the life from the imbiber.
+		S.adjustBruteLoss(1)
+		M.adjustBruteLoss(1)
+	..()
```
return ..()

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -481,18 +542,24 @@
+/datum/reagent/consumable/ethanol/devilskiss/on_mob_life(mob/living/M)
+	for(var/mob/living/simple_animal/slaughter/S in range(10, M)) //Devil's Kiss has the same effect as Demon's Blood, only stronger.
+		S.adjustBruteLoss(2)
+		M.adjustBruteLoss(2)
```
why is there two adjustbruteloss?

GunHog replied:
Seems to deal damage to each slaughter demon in range, as well as the drinker.

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -481,18 +542,24 @@
+	for(var/mob/living/simple_animal/slaughter/S in range(10, M)) //Devil's Kiss has the same effect as Demon's Blood, only stronger.
+		S.adjustBruteLoss(2)
+		M.adjustBruteLoss(2)
+	..()
```
return ..()

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -575,7 +648,12 @@
+/datum/reagent/consumable/ethanol/syndicatebomb/on_mob_life(mob/living/M)
+	if(prob(5))
+		playsound(get_turf(M), 'sound/effects/explosionfar.ogg', 100, 1)
+	..()
```
return ..()

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -561,7 +628,13 @@
+	if(M.mind && M.mind.changeling) //Changeling Sting assists in the recharging of changeling chemicals.
+		M.mind.changeling.chem_charges += metabolization_rate
+		M.mind.changeling.chem_charges = Clamp(M.mind.changeling.chem_charges, 0, M.mind.changeling.chem_storage)
+	..()
```
return ..()

phil235 commented on code/modules/mob/living/carbon/human/human_defines.dm:
```diff
@@ -53,3 +53,8 @@
+	var/drunkenness = 0 //Overall drunkenness - check handle_alcohol() in life.dm for effects
+	var/drunk_state = NOT_DRUNK //A mob's state of drunkenness
+	var/alcohol_tolerance = AVERAGE_TOLERANCE //A mob's tolerance for alcohol poisoning
+	var/hungover = 0 //Whether or not they're recovering from heavy drinking and are still thirsty
```
four new vars for humans? ngggghh

Iamgoofball replied:
tee bee aych this should be on the reagents holder itself like addicton shit is

so monkeys can get hung over too

phil235 commented on code/modules/mob/living/living.dm:
```diff
@@ -505,6 +505,9 @@ Sorry Giacom. Please don't be mad :(
 	fire_stacks = 0
+	if(ishuman(src))
+		var/mob/living/carbon/human/H = src
+		H.modify_drunkenness(-H.drunkenness)
```
why not put this here:https://github.com/tgstation/-tg-station/blob/6949167e6fc674e71988be6713925d8953bb1e2d/code/modules/mob/living/carbon/human/human.dm#L948

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -58,20 +56,35 @@
+/datum/reagent/consumable/ethanol/beer/green/on_mob_life(mob/living/M)
+	if(M.color != color)
+		M.color = color
+	..()
```
return ..()

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -180,35 +193,48 @@
+		var/mob/living/carbon/human/H = M
+		if(H.alcohol_tolerance == initial(H.alcohol_tolerance))
+			H.alcohol_tolerance = BELOW_AVERAGE_TOLERANCE //Hooch is really strong and makes you less tolerant
+	..()
```
return ..()

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -422,30 +470,43 @@
+/datum/reagent/consumable/ethanol/demonsblood/on_mob_life(mob/living/M)
+	for(var/mob/living/simple_animal/slaughter/S in range(10, M)) //Demon's Blood causes nearby slaughter demons to sap the life from the imbiber.
+		S.adjustBruteLoss(1)
+		M.adjustBruteLoss(1)
```
why is there two adjustbruteloss here?

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -481,18 +542,24 @@
+	boozepwr = 70
+
+/datum/reagent/consumable/ethanol/devilskiss/on_mob_life(mob/living/M)
+	for(var/mob/living/simple_animal/slaughter/S in range(10, M)) //Devil's Kiss has the same effect as Demon's Blood, only stronger.
```
range(10) ? isn't this going to be pretty costly?

Xhuis commented:
New in this commit:
* Changed the functionality of Demon's Blood and Devil's Kiss. Demon's Blood prevents you from being dragged into blood pools by demons, and Devil's Kiss causes demons to expel you from the blood pool and take slight damage when trying to eat you
* Only human revival removes drunkenness and resets the drunk state; it is no longer in /mob/living/revive()
* General code optimizations

Xhuis commented:
New in this commit:
* Ported Whiskey Sour and Fetching Fizz from #16134, in addition to the framework for different mixing sounds

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -316,50 +364,49 @@
 	M.Stun(2)
-	..()
-	return 1
+	return ..()
```
no, the previous code was fine, since Stun is always called we must always return 1, it doesn't matter what ..() returns.

Also, the M.Stun(2) above should be `M.Stun(2, 0)`, I must've missed it when I did the change.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -35,6 +35,8 @@
 		handle_heart()
 
+		handle_alcohol() //Alcohol
+
```
I'm not a big fan of having every human call handle_alcohol() every tick when they will only have actual alcohol in their reagent like 1% of the time.
I made a 2k lines change PR simply to remove 2 proc calls in Life() so I'm not a big fan of adding new proc calls to Life().

Xhuis replied:
What do you suggest?

phil235 replied:
maybe move it to where dizziness is handled, handle_status_effect() or something

Zilenan91 commented:
Back in the day, antag testing with syndicakes was bad because /tg/ had a lot of line-toeing assholes who have since then almost all been banned. It wouldn't be a problem anymore to have antag specific healing.

PKPenguin321 commented:
>Back in the day, antag testing with syndicakes was bad because /tg/ had a lot of line-toers who have since then almost all been banned. It wouldn't be a problem anymore to have antag specific healing.

No, it would, but for revs it doesn't really matter. They're very common and you can test for revs by just beating their heads in anyways, no drink required.

Lumipharon commented:
Does this pr mean that you don't become a retardly drunk mess that can't talk good after 2 sips of vodka?
Because seriously, currently you get so fucking drunk, so fucking fast it's not funny.

Xhuis commented:
@Lumipharon I would say that this arguably makes you drunk _faster_, but only if you drink a large amount. Even a couple of sips of crude moonshine won't get you to fatal poisoning.

Lumipharon commented:
The literal speed wasn't the issue, it was the fact that if you sit down at the bar to try and RP, you'd go through a quarter of a glass and you would be this slurring moron.

spess men not being able to hold their liquor is really annoying. Back on NT someone simply toned down the alcohol strengths of most of the drinks and basically fixed this.

Xhuis commented:
I can understand that. I'll see what I can do.

Xhuis commented:
New in this commit:
* Added Absinthe. It is a base alcohol (like whiskey or gin) and can be found in booze dispensers and the Booze-O-Mat. It has a booze power of 80, making it the most powerful base alcohol. It is also rumored to cause hallucinations, but that's probably not true.
* Added Hearty Punch. Five units of absinthe, syndicate bomb, and brave bull are required to make one unit. It has a low metabolization rate and booze power of 10. While the imbiber is in critical condition, they will heal extremely quickly. Upon their death, if they still have the drink in their bloodstream, half of it will be consumed to prevent the death and heal an amount of all damage types equal to the amount of Hearty Punch in their bloodstream. ___Balance changes will be coming to this drink in future commits based on feedback.___
* Passive sobering rate is now affected by alcohol tolerance.
* Probability of vomiting with semi-severe alcohol poisoning decreased to 5% per tick (down from 15% per tick).
* Soda and booze dispensers have been tweaked. Booze dispensers no longer contain any non-booze liquids, but soda dispensers have a wider variety for mixing cocktails.
* Support for custom mixing sounds has been ported from #16134.

Xhuis commented:
New in this commit:
* Hearty Punch is now completely removed upon cheating death. Atropine of an equivalent amount to the Hearty Punch is added to stabilize the patient.

Fox-McCloud commented on code/modules/mob/living/carbon/human/death.dm:
```diff
@@ -16,6 +16,19 @@
+		adjustOxyLoss(-amount_of_punch)
+		modify_drunkenness(80) //There has to be SOME downside
+		reagents.remove_reagent("hearty_punch", amount_of_punch)
+		reagents.add_reagent("atropine", amount_of_punch) //To ensure that they don't immediately bleed out
```
Atropine is nasty as hell; epinephrine would be better.

Xhuis replied:
Hearty Punch is tedious to make and requires basically a full chem dispenser's worth of energy to make even ten units. I just wanted to make a nonconventional cocktail with an actual effect - not to mention if someone is bludgeoning you, they can likely kill you faster than you can heal.

phil235 replied:
needs a if(!gibbed)

Xhuis commented:
New in this commit:
* Sleeping makes recovery from alcohol poisoning over seven times faster.
* Drunkenness states have been moved to misc.dm in __DEFINES.
* Hangovers and alcohol tolerance, both of which were unfinished, have been removed and will likely not be included in the final product.
* Added Bacchus' Blessing. A mixture of four of the most extremely powerful alcohols on the station, it is a revolting mixture that is _extremely_ strong in its alcohol content.

Iamgoofball commented:
I'll work on hangovers and tolerance

Fox-McCloud commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -625,4 +690,48 @@
+/datum/reagent/consumable/ethanol/fetching_fizz/on_mob_life(mob/living/M)
+	for(var/obj/item/weapon/ore/O in orange(3, M))
+		step_towards(O, get_turf(M))
+	..()
```
How a reagent, that has no alcohol in it and has a positive reagent in it that boosts your speed suddenly becomes alcoholic and effectively space drugs, upon mixing with uranium (which nuka cola already has) and iron is beyond me.

Iamgoofball replied:
![100a6a7c2fafbc9e79699fcf584c8e1b](https://cloud.githubusercontent.com/assets/4081722/13868979/166d1a34-ec8c-11e5-882e-762af0cf26f2.jpg)

phil235 replied:
return ..()

Fox-McCloud commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -625,4 +690,48 @@
+		M.adjustToxLoss(-5, 0)
+		M.adjustOxyLoss(-5, 0)
+		. = 1
+	return ..() || .
```
People are going to start using this in cryo mixes in combination with anihol for insanely ridiculous heal speeds. It's not hard to make when all the reagents in this are straight from the dispenser.

Iamgoofball replied:
Okay. And?

Xhuis replied:
Like I've said twice, it only affects people in critical condition, so once they're out of crit it does essentially nothing

As I've _also_ mentioned, while it's easy to make, it's very tedious. It takes _15_ units of ingredients to make 3 units of hearty punch.

duncathan replied:
This is a nice return statement

Fox-McCloud replied:
the `on_mob_life` will occur if they're unconscious--meaning sleeping, paralysis, and the likes will trigger it as well---and a massive 5 heal of each type, at that.

Pressing the same 3 buttons in a row to get this isn't difficult, it's barely even tedious once you get onto it.

I just don't think this is a good idea when cryo is already the quintessential "shove them in, regardless of damage type and heal them back to full--who cares about medical chems".

PKPenguin321 replied:
to be fair cryo is shit right now

KazeEspada replied:
Fox, hearty punch requires heating. And its a 15 to 1 ratio.

Iamgoofball replied:
@Xhuis 
Just switch the uncon. check for a if(M.health > 0) check

that will check for crit

KazeEspada commented on code/modules/food&drinks/recipes/drinks_recipes.dm:
```diff
@@ -577,4 +577,39 @@
+	result = "bacchus_blessing"
+	required_reagents = list("hooch" = 1, "absinthe" = 1, "manlydorf" = 1, "syndicatebomb" = 1)
+	result_amount = 4
+	mix_message = "<span class='warning'>The mixture turns to a sickening froth.</span>"
```
Mix sound should be prayerchaplain

Xhuis commented:
New in this commit:
* Due to runtime errors, Hearty Punch has been changed in functionality. It is now more effective in healing, and also heals clone damage, but cannot prevent death. It is still a 5/5/5 to 1 ratio of ingredients to result.

PKPenguin321 commented:
alcohol is the new trek chems and im loving it

Zilenan91 commented:
Since Hearty Punch can be made totally in the bar could you possibly buff doctor's delight?

phil235 commented on code/modules/mob/living/carbon/human/human_defines.dm:
```diff
@@ -53,3 +53,6 @@
 	var/heart_attack = 0
+
+	var/drunkenness = 0 //Overall drunkenness - check handle_alcohol() in life.dm for effects
+	var/drunk_state = NOT_DRUNK //A mob's state of drunkenness
```
You don't need two variables for this. Just use drunkness and make defines for threshold values.

phil235 commented on code/__DEFINES/misc.dm:
```diff
@@ -378,4 +378,13 @@ var/list/bloody_footprints_cache = list()
+#define HEAVILY_DRUNK 3
+#define SMASHED 4
+#define SHITFACED 5
+#define BARELY_ALIVE 6
```
These defines are a bit too short, they could be conflicting with other defines in the future. And BARELY_ALIVE is way too generic. Use something like ALCOHOL_SOBER, ALCOHOL_TIPSY, ALCOHOL_DRUNK, it's more specific.

Iamgoofball replied:
I highly doubt we're ever gonna use SHITFACED again

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -58,20 +56,35 @@
+		var/mob/living/carbon/human/H = M
+		spawn(300)
+			if(H)
+				H.drunk_state = NOT_DRUNK
```
why is this needed when you have some code changing drunk_state depending on the drunkeness value?

Xhuis replied:
When the alcohol reagent leaves someone's body, the drunk state doesn't ever reset. I'll optimize it in the next commit.

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -19,36 +8,45 @@
-			. = 1
+	if(ishuman(M))
+		var/mob/living/carbon/human/H = M
+		H.modify_drunkenness(boozepwr * 0.01)
```
You could give an argument to modify_drunkenness based on boozepwr to limit the maximum value reachable with a given type of alcohol. Just an idea.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,88 @@
+91-100: Dangerously toxic - swift death
+*/
+
+/mob/living/carbon/human/proc/handle_alcohol()
```
Can you at least start your proc with if(drunkenness) , so every single mob doesn't have to go through 11 IFs every single tick????

Xhuis replied:
Sure, that'll work.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,88 @@
+				sleeping += 45
+
+	if(drunkenness >= 101)
+		adjustToxLoss(4) //Let's be honest you shouldn't be alive by now
```
why not use a switch(drunkeness) ?

Xhuis replied:
I tried that and it didn't want to work. My question to Remie about how effects could be inherited got no answers.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,88 @@
+		drunk_state = BARELY_ALIVE
+		if(prob(20))
+			if(SSshuttle.emergency.mode == SHUTTLE_DOCKED && z == ZLEVEL_STATION) //QoL mainly
+				src << "<span class='warning'>You're so tired... but you can't miss that shuttle...</span>"
```
Hacky. Also, why should we give drunk people this chance when all other things that can make you sleep do not?

Xhuis replied:
I knew that people would whine about passing out because they were about to board the shuttle and get greentext and then passed out from RNG. I could remove it if you want me to.

phil235 replied:
Let them whine, people don't get worse than shifaced by accident.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,88 @@
+				src << "<span class='warning'>You're so tired... but you can't miss that shuttle...</span>"
+			else if(!sleeping)
+				src << "<span class='warning'>Just a quick nap...</span>"
+				sleeping += 45
```
You must use Sleeping(45) , manually changing the sleeping var does not put the mob to sleep.

Xhuis replied:
It _does_ work, but it has a delay. I'll change that.

phil235 replied:
A random stat update can make it work but it still doesn't change the sleeping alert.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,88 @@
+		jitteriness = max(jitteriness - 3, 0)
+
+	if(drunkenness >= 11)
+		slurring += 1.2 //Constant but only a very slow increase
```
Maybe you should add a max, we don't want someone who stay slightly tipsy for a long time to slur for the rest of the shift.

Xhuis replied:
Good point.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,88 @@
+
+	if(drunkenness >= 1)
+		if(prob(25))
+			slurring += 2
```
starting to slur and jitter after drinking a single drop of beer??

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,88 @@
+		drunk_state = SMASHED
+
+	if(drunkenness >= 71)
+		adjust_blurriness(1.2)
```
I think blur_eyes(5) is more appropriate here, we don't want the blurriness to accumulate and stay for a very long time if someone get drunk for a long time and sobers up.

Xhuis replied:
All right.

Bawhoppen commented:
I didn't bother to read everything but I still want all alcoholic beverages to be usable as ethanol in making chems for some clowning fun

Xhuis commented:
New in this commit:
* drunk_state and its defines have been removed. Examining for drunkenness now checks for the drunkenness variable instead.
* Slurring can now only stack to 5 as a result of alcohol poisoning.
* blur_eyes() has replaced adjust_blurriness(), and Sleeping() is now properly used.
* Alcohol poisoning rate has been increased by 50%.
* Alcohol poisoning speed is now controlled by a define, ALCOHOL_RATE, in alcohol_reagents.dm.

Commit squash will be coming soon.

Commit:
Removes uranium from Fetching Fizz recipe

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -422,30 +463,45 @@
 
-/datum/reagent/consumable/ethanol/demonsblood
-	name = "Demons Blood"
+/datum/reagent/consumable/ethanol/demonsblood //Prevents the imbiber from being dagged into a pool of blood by a slaughter demon.
```
"dragged"

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -316,50 +358,49 @@
+	boozepwr = 70
 
-/datum/reagent/consumable/ethanol/manly_dorf
+/datum/reagent/consumable/ethanol/manly_dorf //THE SPRITE FOR THIS HAS A HANDLE AND THE HANDLE HAD A STRIP OF WHITE PIXELS BETWEEN IT AND THE MUG AND I FINALLY REMOVED IT AND OH GOD I'M SO HAPPY
```
I don't get the comment...

Xhuis replied:
The current sprite for Manly Dorf is a mug with a handle. Between the handle and mug is a block of white pixels that has irked me endlessly. In this PR I removed the block of white pixels to make it look more like an actual cup.

phil235 replied:
We don't need comments like that. It doesn't inform the reader.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,83 @@
+		modify_drunkenness(-0.2) //Becoming sober takes much longer than becoming drunk
+
+	if(!drunkenness)
+		return 0
```
put this at the beginning of the proc, not after the sleeping check and modify_drunkenness() call.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,83 @@
+	if(drunkenness >= 101)
+		adjustToxLoss(4) //Let's be honest you shouldn't be alive by now
+
+/mob/living/carbon/human/proc/modify_drunkenness(var/amount)
```
we don't need the "var/" in front of the argument.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,83 @@
+	if(sleeping)
+		if(drunkenness >= 101)
+			drunkenness -= 25 //Mainly to ensure that you don't goddamn die from it for balance reasons
+	drunkenness = Clamp(drunkenness, 0, INFINITY)
```
I don't like this proc to be honest, the only reason you use it is because of this sleeping thing which is a hacky way to handle a balance problem.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,83 @@
+	if(drunkenness >= 81)
+		adjustToxLoss(0.2)
+		if(prob(5))
+			src << "<span class='warning'>Maybe you should lie down for a bit...</span>"
```
You're going to get these messages even if you're unconscious.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -318,4 +320,83 @@
+	if(drunkenness >= 6)
+		if(prob(25))
+			slurring += 2
+		jitteriness = max(jitteriness - 3, 0)
```
So we need to be tipsy to get the jitteriness reduction effect? tbh I don't know whether this is more or less realistic than the old effect :shrug:

Commit:
Code optimization

Xhuis commented:
New in this commit:
* Examining for drunkenness has been made more foolproof.
* handle_alcohol() has been migrated to handle_status_effects() for humans and now only occurs if the user is drunk.
* Brain loss now occurs in small amounts when near the highest level of alcohol poisoning.
* modify_drunkenness() has been optimized.

phil235 commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -250,15 +250,15 @@
 			if(11 to 21)
 				msg += "[t_He] [t_is] slightly flushed.\n"
-			if(22 to 41)
+			if(21.01 to 41) //.01s are used in case drunkenness ends up to be a small decimal
```
why not 21 ?

Xhuis replied:
I thought that would make the two categories conflict. If it doesn't, I'll fix it.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -335,68 +333,61 @@ All effects don't start immediately, but rather get worse over time; the rate is
+
+/mob/living/carbon/human/proc/modify_drunkenness(amount)
+	if(sleeping)
+		amount *= 5
```
this would mean you'd get more drunk is asleep with alcohol in your reagents.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -335,68 +333,61 @@ All effects don't start immediately, but rather get worse over time; the rate is
+		if(drunkenness >= 81)
+			adjustToxLoss(0.2)
+			if(prob(5) && !sleeping)
+				src << "<span class='warning'>Maybe you should lie down for a bit...</span>"
```
you'll still get those messages while in crit though.

phil235 commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -335,68 +333,61 @@ All effects don't start immediately, but rather get worse over time; the rate is
+/mob/living/carbon/human/handle_status_effects()
+	..()
+	if(drunkenness)
+		modify_drunkenness(-0.2) //Slow sobering - sleeping increases it to 1 per tick
```
if(drunkenness)
	if(sleeping)
		drunkenness = max(drunkenness - alot, 0)
	else
		drunkenness = max(drunkenness - 0.2, 0)

Commit:
Implements @phil235's recommended changes

Commit:
Fun removal

phil235 commented:
You left a modify_drunkenness in human.dm

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -1,17 +1,8 @@
+#define ALCOHOL_RATE 0.015 //The rate at which alcohol affects you; by default, one glass of whiskey gets you just above tipsy
```
gets you just above what? drunk level? shitfaced level?

Also, are you sure about your calculations? With whiskey I get: 0.015 * 75 = 1.125, minus the 0.2 drunkenness reduction per tick, you get +0.925 per tick of whiskey consumed. Since a glass is 50u and you metabolize 0.4u per tick, a full glass of whiskey makes you reach drunkenness = 115

Xhuis replied:
It was just an assumption. I'll change the comment

KorPhaeron commented on code/modules/mob/living/carbon/human/death.dm:
```diff
@@ -16,6 +16,7 @@
 /mob/living/carbon/human/death(gibbed)
 	if(stat == DEAD)
 		return
+
```
?

Xhuis replied:
Left over from when Hearty Punch prevented death. I can remove it if it's that bad.

KorPhaeron commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -246,6 +246,21 @@
 	if(reagents.has_reagent("teslium"))
 		msg += "[t_He] is emitting a gentle blue glow!\n"
 
+	if(drunkenness && !skipface && stat != DEAD) //Drunkenness
```
Do we really not have a better way to add to examines than a series of endless if statements?

Xhuis replied:
Probably not.

Commit:
A few more optimizations

phil235 commented on code/modules/reagents/chemistry/reagents/alcohol_reagents.dm:
```diff
@@ -19,36 +10,45 @@
-			. = 1
+	if(ishuman(M))
+		var/mob/living/carbon/human/H = M
+		H.drunkenness = max((H.drunkenness + boozepwr * ALCOHOL_RATE), 0)
```
I'm still a bit concerned about balance. A single full glass of whiskey giving you brain damage is a bit too much.

The model isn't realistic, any alcohol with boozepwr below 14 can never make you even tipsy no matter how much you drink. On the other side of the spectrum, drinking small sips of whisky slowly but regularly will have you ending up beyond shitfaced.

It might be a good idea to include the reagent volume in the drunkenness calculation, maybe make it use a Clamp so your drunkenness can never go above a certain value.

Xhuis replied:
I've been a bit torn on what formula to use as well. I'll have to think of something.

Xhuis replied:
I've gotten a headache trying out different formulas. Any ideas?

phil235 replied:
Something like this maybe:
```
if(H.drunkenness < volume * boozepwr * 0.05)
	H.drunkenness = H.drunkenness + sqrt(volume) * boozepwr * 0.005
```

Xhuis replied:
I'll test it out when I get home. Thanks!

duncathan commented:
I hate when people make gameplay changes in a pr that's supposedly a refactor

Xhuis commented:
Could you possibly be any more passive-aggressive?

Commit:
Modifies alcohol poisoning formula, thanks to @phil235

Xhuis commented:
Thanks again for the formula. It's working a lot better than the one I had dreamt up - my only gripe with it is that it generates really long decimals, but that's just a nitpick.

phil235 commented:
You can add a round( xxx, 0.1) to reduce the number of decimals to 1.

Xhuis commented:
That doesn't concern me, as the switch has ways to handle it (in examines). This should be merge-ready.
